contracts/pool/pools/liquid/WithdrawalEligibility.sol

```diff
@@ -0,0 +1,155 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.18;
+
+import "../TokenPoolBase.sol";
+
+error UnlockPeriodNotFinished();
+error NonWithdrawableTokens();
+error AmountExceedsAvailableWithdraw();
+
+/**
+ * @title WithdrawalEligibility
+ * @dev Provides functionality for checking withdrawal eligibility and grace periods
+ */
+contract WithdrawalEligibility is TokenPoolBase {
+    uint256 internal constant GRACE_PERIOD = 4 weeks;
+    uint256 internal constant GRACE_PERIOD_REMOVAL_FEE = 1000;
+    uint256 internal constant UPGRADE_PERIOD = 63_158_400; // 2 years considering leap years in seconds;
+
+    struct EligbleWithdraw {
+        uint256 gracePeriodEndTime;
+        uint256 amount;
+    }
+
+    struct Stake {
+        EligbleWithdraw eligbleWithdraw;
+        uint256 amount;
+    }
+
+    struct Lock {
+        EligbleWithdraw eligbleWithdraw;
+        uint256 lockTime;
+        uint256 amount;
+    }
+
+    mapping(address => Stake) internal stakes;
+
+    mapping(address => Lock) internal locks;
+
+    event GracePeriodStarted(
+        address indexed account,
+        uint256 indexed unlockPeriodEnd
+    );
+
+    function startWithdrawPeriod(
+        uint256 amount
+    ) external onlySufficientAmount(amount) {
+        Stake memory stakeInfo = stakes[msg.sender];
+
+        _startGracePeriod(stakeInfo.eligbleWithdraw, amount);
+
+        stakes[msg.sender] = stakeInfo;
+
+        emit GracePeriodStarted(msg.sender, stakeInfo.eligbleWithdraw.amount);
+    }
+
+    function startUnlockPeriod(
+        uint256 amount
+    ) external onlySufficientAmount(amount) {
+        Lock memory lockInfo = locks[msg.sender];
+
+        _startGracePeriod(lockInfo.eligbleWithdraw, amount);
+
+        locks[msg.sender] = lockInfo;
+
+        emit GracePeriodStarted(
+            msg.sender,
+            lockInfo.eligbleWithdraw.gracePeriodEndTime
+        );
+    }
+
+    function _startGracePeriod(
+        EligbleWithdraw memory eligbleWithdraw,
+        uint256 amount
+    ) internal view {
+        _checkWithdrawEligibility(amount);
```

Comment:
Ova provjera da je `amount` razliciti od nule mislim da je redundantna jer to vec provjeravamo s modifierom `onlySufficientAmount` kad pozivamo `startWithdrawPeriod` i `startUnlockPeriod` funkcije

Reply:
slazem se dobro si uhvatia